locator/token_metadata.hh

```diff
@@ -390,4 +394,25 @@ public:
     static future<> mutate_on_all_shards(sharded<shared_token_metadata>& stm, seastar::noncopyable_function<future<> (token_metadata&)> func);
 };
 
-}
+} // namespace locator
+
+namespace std {
+
+template<>
+struct hash<locator::host_id_or_endpoint> {
+    size_t operator()(const locator::host_id_or_endpoint& i) const {
+        return i.has_host_id() ? hash<locator::host_id>()(i.id) : hash<gms::inet_address>()(i.endpoint);
+    }
+};
+
+} // namespace std
+
+template <>
+struct fmt::formatter<locator::host_id_or_endpoint> {
+    constexpr auto parse(format_parse_context& ctx) { return ctx.begin(); }
+
+    template <typename FormatContext>
+    auto format(const locator::host_id_or_endpoint& hioa, FormatContext& ctx) const {
+        return fmt::format_to(ctx.out(), "{}/{}", hioa.id, hioa.endpoint);
```

Comment:
If it's host_id OR endpoint, then we should print either id or endpoint.

Reply:
It's an inclusive OR.  Both can be present.
Otherwise, a std::variant could be used as well.  Printing both would allow us to see the full state.
We can becsmarter and print only one if them if the other is missing, or both in case both are present.